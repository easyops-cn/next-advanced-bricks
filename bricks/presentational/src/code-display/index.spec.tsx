import { describe, test, expect, jest } from "@jest/globals";
import { act } from "react-dom/test-utils";
import { fireEvent } from "@testing-library/react";
import ResizeObserver from "resize-observer-polyfill";
import fileSaver from "file-saver";
import "./";
import { CodeDisplay } from "./index.js";

jest.mock("@next-core/theme", () => ({}));

jest.mock("prismjs/components/prism-markdown.min.js", () => {
  throw new Error();
});

jest.useFakeTimers();

const consoleError = jest.spyOn(console, "error").mockReturnValue();

const spyOnSaveAs = jest.spyOn(fileSaver, "saveAs").mockImplementation(() => {
  // do nothing
});

const copyToClipboard = jest.fn().mockResolvedValue(1 as never);
const showNotification = jest.fn();

customElements.define(
  "basic.copy-to-clipboard",
  class extends HTMLElement {
    resolve = copyToClipboard;
  }
);
customElements.define(
  "basic.show-notification",
  class extends HTMLElement {
    resolve = showNotification;
  }
);
global.ResizeObserver = ResizeObserver as any;

describe("eo-code-display", () => {
  test("basic usage", async () => {
    const element = document.createElement("eo-code-display") as CodeDisplay;
    element.language = "javascript";
    element.value = "const a = 1;";
    element.showCopyButton = true;
    element.showExportButton = true;
    element.maxLines = 10;
    element.minLines = 3;
    element.exportFileName = "export.js";

    expect(element.shadowRoot).toBeFalsy();

    await act(() => {
      document.body.appendChild(element);
    });
    expect(element.shadowRoot?.childNodes.length).toBeGreaterThan(1);

    expect(element.shadowRoot?.querySelector("pre")).toMatchSnapshot();
    const pre = element.shadowRoot?.querySelector("pre") as HTMLPreElement;
    expect(pre.classList.contains("line-numbers")).toBeTruthy();
    expect(pre.style.getPropertyValue("--min-lines")).toBe("3");
    expect(pre.style.getPropertyValue("--max-lines")).toBe("10");

    act(() => {
      fireEvent.click(element.shadowRoot?.querySelector(".copy-icon"));
    });
    await (global as any).flushPromises();
    expect(copyToClipboard).toHaveBeenLastCalledWith("const a = 1;");
    expect(showNotification).toHaveBeenLastCalledWith(
      expect.objectContaining({ type: "success" })
    );

    act(() => {
      fireEvent.click(element.shadowRoot?.querySelector(".export-icon"));
    });
    expect(spyOnSaveAs).toHaveBeenLastCalledWith(expect.any(Blob), "export.js");

    act(() => {
      document.body.removeChild(element);
    });
    expect(element.shadowRoot?.childNodes.length).toBe(0);
  });

  test("load language", async () => {
    const element = document.createElement("eo-code-display") as CodeDisplay;

    // base set of languages
    await act(async () => {
      element.language = "javascript";
      document.body.appendChild(element);
    });

    // not in the base set of languages
    await act(async () => {
      element.language = "rust";
    });

    // unsupported language
    await act(async () => {
      element.language = "__unsupported__";
    });
    expect(consoleError).toHaveBeenLastCalledWith(
      expect.stringContaining("unsupported language: __unsupported__")
    );

    // load language failed
    await act(async () => {
      element.language = "markdown";
    });
    expect(consoleError).toHaveBeenLastCalledWith(
      expect.stringContaining("load language failed: markdown")
    );

    act(() => {
      document.body.removeChild(element);
    });
  });
});
