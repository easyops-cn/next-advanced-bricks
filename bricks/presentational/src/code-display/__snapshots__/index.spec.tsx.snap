// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`eo-code-display basic usage 1`] = `
<pre
  class="prism-code language-javascript line-numbers"
  style="color: rgb(204, 204, 204); background: rgb(45, 45, 45); font-family: Consolas, Monaco, "Andale Mono", "Ubuntu Mono", monospace; tab-size: 4; hyphens: none; --max-lines: 10; --min-lines: 3;"
>
  <div
    class="token-line"
  >
    <span
      class="line-number"
    >
      1
    </span>
    <span>
      <span
        class="token keyword"
        style="color: rgb(204, 153, 205);"
      >
        const
      </span>
      <span
        class="token plain"
      >
         a 
      </span>
      <span
        class="token operator"
        style="color: rgb(103, 205, 204);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token number"
        style="color: rgb(240, 141, 73);"
      >
        1
      </span>
      <span
        class="token punctuation"
        style="color: rgb(204, 204, 204);"
      >
        ;
      </span>
    </span>
  </div>
</pre>
`;
